src/hotspot/share/interpreter/bytecodeTracer.cpp

```diff
@@ -141,64 +148,45 @@ class BytecodePrinter: public BytecodeClosure {
     _code = code;
     int bci = bcp - method->code_base();
     // Print bytecode index and name
+    if (ClassPrinter::has_mode(_flags, ClassPrinter::PRINT_BYTECODE_ADDR)) {
+      st->print(INTPTR_FORMAT " ", p2i(bcp));
+    }
     if (is_wide()) {
-      st->print("%d %s_w", bci, Bytecodes::name(code));
+      st->print("%4d %s_w", bci, Bytecodes::name(code));
     } else {
-      st->print("%d %s", bci, Bytecodes::name(code));
+      st->print("%4d %s", bci, Bytecodes::name(code));
     }
     _next_pc = is_wide() ? bcp+2 : bcp+1;
     print_attributes(bci, st);
     bytecode_epilog(bci, st);
   }
 };
 
+// We need a global instance to keep track of the states when the bytecodes
+// are executed. Access by multiple threads are controlled by ttyLocker.
+static BytecodePrinter _interpreter_printer;
 
-// Implementation of BytecodeTracer
-
-// %%% This set_closure thing seems overly general, given that
-// nobody uses it.  Also, if BytecodePrinter weren't hidden
-// then Method* could use instances of it directly and it
-// would be easier to remove races on _current_method and bcp.
-// Since this is not product functionality, we can defer cleanup.
-
-BytecodeClosure* BytecodeTracer::_closure = NULL;
-
-static BytecodePrinter std_closure;
-BytecodeClosure* BytecodeTracer::std_closure() {
-  return &::std_closure;
-}
-
-
-void BytecodeTracer::trace(const methodHandle& method, address bcp, uintptr_t tos, uintptr_t tos2, outputStream* st) {
-  if (_closure == NULL) {
-    return;
-  }
-
+void BytecodeTracer::trace_interpreter(const methodHandle& method, address bcp, uintptr_t tos, uintptr_t tos2, outputStream* st) {
   if (TraceBytecodes && BytecodeCounter::counter_value() >= TraceBytecodesAt) {
     ttyLocker ttyl;  // 5065316: keep the following output coherent
```

Comment:
We should file a a RFE to make this a leaf mutex rather than ttyLocker (which does get broken to allow safepoints!) with no_safepoint_check assuming that the printing doesn't safepoint.  Good to remove the no-longer accurate comment.

Reply:
The comment also says:

```
    // Using the ttyLocker prevents the system from coming to
    // a safepoint within this code, which is sensitive to Method*
    // movement.
```

So are we trying to "prevent Method movement"? Does this even make sense after we switched to Permgen?

Reply:
Nope doesn't make sense.

Reply:
I filed https://bugs.openjdk.org/browse/JDK-8295851 "Do not use ttyLock in BytecodeTracer::trace"